### PR TITLE
Aarch64 task running and register logic

### PR DESCRIPTION
**Pull request by Keno**
As discussed in #2564, aarch64 register definition is a bit tricky,
because the kernel does not give us a way to request orig_x0, so we
need to track it ourselves. This means splitting the register format
in the trace from just a default Arch::user_regs_struct to include
the extra orig_x0. Additionally, we need to make sure to copy x0 to
this field on every syscall entry.

### CODE REVIEW

File: src/AutoRemoteSyscalls.cc [hunk 1]
```diff
@@ -255,6 +256,16 @@ long AutoRemoteSyscalls::syscall_base(int syscallno, Registers& callregs) {
         // We entered the syscall, so stop now
         break;
       }
+      if (t->ptrace_event() == PTRACE_EVENT_EXIT) {
+        // We died, just let it be
+        break;
+      }
+      if (t->stop_sig() == SIGTRAP && t->get_siginfo().si_code == TRAP_TRACE) {
```
Comment (khuey) [inline src/AutoRemoteSyscalls.cc:263]: If this is AArch64 specific, perhaps it should be ifdefed?

File: src/AutoRemoteSyscalls.cc [hunk 2]
```diff
@@ -255,6 +256,16 @@ long AutoRemoteSyscalls::syscall_base(int syscallno, Registers& callregs) {
         // We entered the syscall, so stop now
         break;
       }
+      if (t->ptrace_event() == PTRACE_EVENT_EXIT) {
+        // We died, just let it be
```
Comment (khuey) [inline src/AutoRemoteSyscalls.cc:260]: Is this a theoretical concern or did you actually see it happen?

File: src/Monkeypatcher.cc [hunk 3]
```diff
@@ -937,7 +937,22 @@ void patch_after_exec_arch<X64Arch>(RecordTask* t, Monkeypatcher& patcher) {
...
+
+  for (const auto& m : t->vm()->maps()) {
+    auto& km = m.map;
+    patcher.patch_after_mmap(t, km.start(), km.size(),
+                             km.file_offset_bytes()/page_size(), -1,
+                             Monkeypatcher::MMAP_EXEC);
+  }
+
+  if (!t->vm()->has_vdso()) {
+    patch_auxv_vdso(t);
```
Comment (khuey) [inline src/Monkeypatcher.cc:952]: I'm confused, why is this in a !has_vdso branch?

File: src/Registers.h [hunk 4]
```diff
@@ -33,6 +33,8 @@ const uintptr_t X86_DF_FLAG = 1 << 10;
 const uintptr_t X86_RF_FLAG = 1 << 16;
 const uintptr_t X86_ID_FLAG = 1 << 21;
 
+const uintptr_t AARCH64_DBG_SPSR_SS = 1 << 21;
```
Comment (khuey) [inline src/Registers.h:36]: I'm just assuming this is correct because it's shockingly difficult to find what the layout of the PSTATE is in the ARM Reference Manual.

File: src/Task.cc [hunk 5]
```diff
@@ -1085,54 +1089,62 @@ static bool is_singlestep_resume(ResumeRequest request) {
...
-        // This also takes care of the did_set_breakpoint_after_cpuid workaround case
-      } else if (ti == TrappedInstruction::INT3 &&
-          ip() == address_of_last_execution_resume +
-                       trapped_instruction_len(TrappedInstruction::INT3)) {
-        // INT3 instructions should also be turned into a singlestep here.
+  const siginfo_t& si = get_siginfo();
+  if (arch() == x86 || arch() == x86_64) {
+    uintptr_t status = x86_debug_status();
+    reasons.singlestep = (status & DS_SINGLESTEP) != 0;
+    if (!reasons.singlestep && is_singlestep_resume(how_last_execution_resumed)) {
```
Comment (khuey) [inline src/Task.cc:1099]: Why did you add the `!reasons.singlestep` condition?

File: src/Task.cc [hunk 6]
```diff
@@ -1234,24 +1251,25 @@ static bool cpu_has_KNL_string_singlestep_bug() {
...
-    LOG(debug) << "Working around KNL single-step hardware bug (cx=" << cx
-               << ")";
-    if (cx > cutoff) {
-      last_resume_orig_cx = cx;
-      Registers r = regs();
-      /* An arbitrary value < cutoff would work fine here, except 1, since
-         the last iteration of the loop behaves differently */
-      r.set_cx(cutoff);
-      set_regs(r);
+  if (cpu_has_KNL_string_singlestep_bug()) {
```
Comment (khuey) [inline src/Task.cc:1256]: I think it'd be easier just to `if (!cpu_has_KNL_string_singlestep_bug()) { return; }` at the beginning and leave the rest of this function undisturbed.

File: src/Task.cc [hunk 7]
```diff
@@ -1345,13 +1366,35 @@ void Task::flush_regs() {
...
+  if (orig_syscallno_dirty) {
+    uintptr_t syscall = registers.original_syscallno();
+    struct iovec vec = { &syscall,
+                          sizeof(syscall) };
+    LOG(debug) << "Chaning syscall to " << syscall;
+    if (ptrace_if_alive(PTRACE_SETREGSET, NT_ARM_SYSTEM_CALL, &vec)) {
+      orig_syscallno_dirty = false;
     }
   }
+#endif
```
Comment (khuey) [inline src/Task.cc:1397]:
```
#elif defined(__i386__) || defined(__x86_64__)
  ASSERT(t, !orig_syscallno_dirty);
#endif
```

File: src/Task.cc [hunk 8]
```diff
@@ -1345,13 +1366,35 @@ void Task::flush_regs() {
   if (registers_dirty) {
     LOG(debug) << "Flushing registers for tid " << tid << " " << registers;
     auto ptrace_regs = registers.get_ptrace();
+#if defined(__i386__) || defined(__x86_64)
```
Comment (khuey) [inline src/Task.cc:1369]: It seems your `__x86_64__` got its right arm torn off in some sort of terrible industrial accident.

File: src/Task.h [hunk 9]
```diff
@@ -454,20 +454,27 @@ class Task {
...
   /**
-   * Set the debug status (DR6 on x86).
+   * Set the debug status (DR6 on x86). Noop on non-x86 architectures.
    */
-  void set_debug_status(uintptr_t status);
+  void set_x86_debug_status(uintptr_t status);
 
   /**
-   * Determine why a SIGTRAP occurred. Uses debug_status() but doesn't
+   * Determine why a SIGTRAP occurred. Ont x86, uses x86_debug_status() but doesn't
```
Comment (khuey) [inline src/Task.h:467]: On

File: src/Task.h [hunk 10]
```diff
@@ -1101,6 +1108,9 @@ class Task {
   // True when 'registers' has changes that haven't been flushed back to the
   // task yet.
   bool registers_dirty;
+  // True when changes to the original syscallno in 'registers' have not been
+  // flushed back to the task yet.
```
Comment (khuey) [inline src/Task.h:1112]: Add "Some architectures (e.g. AArch64) require a separate ptrace call for this."

File: src/util.h [hunk 11]
```diff
@@ -341,7 +341,7 @@ inline bool is_kernel_trap(int si_code) {
   /* XXX unable to find docs on which of these "should" be
    * right.  The SI_KERNEL code is seen in the int3 test, so we
    * at least need to handle that. */
-  return si_code == TRAP_BRKPT || si_code == SI_KERNEL;
+  return si_code == TRAP_TRACE || si_code == TRAP_BRKPT || si_code == SI_KERNEL;
```
[the remark this answers is not in the document]
Comment (Keno) [reply src/util.h:344]: Yes, TRAP_TRACE is single-stepping.

File: src/Monkeypatcher.cc (hunk 3 above)
Comment (Keno) [reply src/Monkeypatcher.cc:952]: this patches out the vdso address from auxv, so that the application doesn't see it. Perhaps the function needs a better name.

File: src/Registers.h (hunk 4 above)
Comment (Keno) [reply src/Registers.h:36]: > it's shockingly difficult to find what the layout of the PSTATE is in the ARM Reference Manual.

right? And yes, I think it's correct.

File: src/AutoRemoteSyscalls.cc (hunk 2 above)
Comment (Keno) [reply src/AutoRemoteSyscalls.cc:260]: I saw it happen, but only due to an unrelated bug that caused the tracee to segfault on the first instruction out of a clone taking down the process before rr was done doing its setup. It's not aarch64 specific.

File: src/remote_code_ptr.h [hunk 12]
```diff
@@ -50,6 +50,18 @@ class remote_code_ptr {
   remote_code_ptr decrement_by_bkpt_insn_length(SupportedArch arch) const {
     return remote_code_ptr(ptr - rr::bkpt_instruction_length(arch));
   }
+  remote_code_ptr undo_executed_bkpt(SupportedArch arch) {
+    switch (arch) {
+      case x86:
+      case x86_64:
+        return decrement_by_bkpt_insn_length(arch);
```
[the remark this answers is not in the document]
Comment (Keno) [reply src/remote_code_ptr.h:57]: Sure.

File: src/Monkeypatcher.cc (hunk 3 above)
Comment (khuey) [reply src/Monkeypatcher.cc:952]: Lets file an issue on renaming it to `patch_auxv_to_remove_vdso` or something.

File: src/Task.cc (hunk 5 above)
Comment (Keno) [reply src/Task.cc:1099]: The only thing that could happen in any of these branches is setting singlestep to `true`, so it isn't necessary to try any of these things if the kernel already told us it was a single step. I originally had the condition inverted to avoid going through some x86 code paths on aarch64, but then I decided to do it this way.

File: src/Task.cc (hunk 8 above)
Comment (Keno) [reply src/Task.cc:1369]: Indeed. No idea why this ever passed tests

File: src/Task.cc (hunk 7 above)
Comment (Keno) [reply src/Task.cc:1397]: We can't really do that, right, because the ptrace above is allowed to fail?
At most we could do:
```
#elif defined(__i386__) || defined(__x86_64__)
  ASSERT(t, registers_dirty == orig_syscallno_dirty);
#endif
```

File: src/AutoRemoteSyscalls.cc (hunk 1 above)
Comment (Keno) [reply src/AutoRemoteSyscalls.cc:263]: I think powerpc has the same behavior, so it might just be x86 that's the odd one out. I don't think It'll do any harm. That said, I am actually a bit confused why TRAP_TRACE doesn't seem to show up for us on x86, because it seems like it should:
```
static inline int get_si_code(unsigned long condition)
{
	if (condition & DR_STEP)
		return TRAP_TRACE;
	else if (condition & (DR_TRAP0|DR_TRAP1|DR_TRAP2|DR_TRAP3))
		return TRAP_HWBKPT;
	else
		return TRAP_BRKPT;
}
[snip]
	si_code = get_si_code(tsk->thread.debugreg6);
	if (tsk->thread.debugreg6 & (DR_STEP | DR_TRAP_BITS) || user_icebp)
		send_sigtrap(regs, error_code, si_code);
```
I do know why we don't see it here, which is that aarch64 reports an extra single step trap after exiting the syscall, but not sure why we don't see it in general (or maybe we do, and I just never looked closely enough).

File: src/Task.cc (hunk 7 above)
Comment (khuey) [reply src/Task.cc:1397]: Ah, yes, good point.

File: src/Task.h (hunk 9 above)
Comment (Keno) [reply src/Task.h:467]: Thanks

File: src/Task.h (hunk 10 above)
Comment (Keno) [reply src/Task.h:1112]: Will do.

File: src/util.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (Keno) [reply src/util.h:344]: See my comment below. I don't actually know why it's not consistent.

**Comment by Keno** [comment]
Updated.

**Comment by Keno** [comment]
Hold on on merging, there was an intermittent test failure that came up while I tested it that looks like it may have been caused by this. One sec while I look into it.

File: src/Task.cc (hunk 7 above)
Comment (Keno) [reply src/Task.cc:1400]: Actually, that's not quite right either, because it is possible for the registers to be dirty, but orig_syscallno not to be dirty. I think the most we can do is maybe something like:
```
  ASSERT(this, !orig_syscallno_dirty || registers_dirty);
```

Comment (khuey) [reply src/Task.cc:1400]: maybe in the x86 `if (registers_dirty)` block add `else { ASSERT(this, !orig_syscallno_dirty); }` then?

Comment (Keno) [reply src/Task.cc:1400]: yep

**Comment by Keno** [comment]
Alright, fixed with the change to the assertion as discussed.

**Comment by khuey** [comment]
@Keno the intermittent you mentioned was that assertion?

**Comment by Keno** [comment]
Yes
